File: 100.相同的树.rs
/*
 * @lc app=leetcode.cn id=100 lang=rust
 *
 * [100] 相同的树
 */

// @lc code=start
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::cell::RefCell;
use std::rc::Rc;
impl Solution {
    pub fn is_same_tree(
        p: Option<Rc<RefCell<TreeNode>>>,
        q: Option<Rc<RefCell<TreeNode>>>,
    ) -> bool {
        match (p, q) {
            (None, None) => true,
            (Some(node1), Some(node2)) => {
                node1.borrow().val == node2.borrow().val
                    && Solution::is_same_tree(
                        node1.borrow().left.clone(),
                        node2.borrow().left.clone(),
                    )
                    && Solution::is_same_tree(
                        node1.borrow().right.clone(),
                        node2.borrow().right.clone(),
                    )
            }
            _ => false,
        }
    }
}
// @lc code=end

File: leetcode/100.相同的树.ts
/*
 * @lc app=leetcode.cn id=100 lang=typescript
 *
 * [100] 相同的树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function isSameTree(p: TreeNode | null, q: TreeNode | null): boolean {
    return dfs(p, q);
};
function dfs(x: TreeNode | null, y: TreeNode | null): boolean {
    if (x === null && y === null) return true;
    if (x === null || y === null) return false;
    return x.val === y.val && dfs(x.left, y.left) && dfs(x.right, y.right);
}
// @lc code=end

